### DISCUSSION TITLE
File based routing groups and index renaming

### DISCUSSION THREAD

**Discussion by robertmoura** [posva/unplugin-vue-router#393] [Ideas]
I really like what solid-start is doing with their file based routing. They've got a concept of renaming index files using `()`. So `/nested/index.vue` could be `/nested/(child).vue` instead. Any chance that something like this might make it into the file based routing?

https://docs.solidjs.com/solid-start/building-your-application/routing#renaming-index

# Update
Posva has added this to the board. Track it at https://github.com/posva/unplugin-vue-router/issues/448.

**Comment by posva** [comment]
In the docs you sent the `(child)` is to avoid nested. You can already achieve that with `nested.child.vue` -> `nested/child`.
They seem to allow `social/index` -> `social/social` but that doesn't help with the problem they mention: searching. When searching, you still need to type social to find the index file but now it's duplicated in the filepath, which can lead to weird situations when debugging

**Reply by robertmoura** [reply]
Yeah you're right 😊 The docs don't seem to mention the index routing. I did notice that but I don't think the docs are completed yet. I've linked a blog post about SolidStart's v1.0 release that mentions the index renaming. It has a good image of what a route structure might look like as well. Here's a quote:

> However, we also use () as a way to do more powerful things like Route Groups, named index routes, and as a way to break out of nesting. Basically whatever you put between the parenthesis is removed from matching, but still impacts the structure of your routes.

![image](https://github.com/posva/unplugin-vue-router/assets/17804984/394a8d97-8acd-42e0-afa2-a04a6e84a936)

https://www.solidjs.com/blog/solid-start-the-shape-frameworks-to-come

About the `social/social` example though, I agree. I'm not a big fan of that idea because it's not as explicit.

**Reply by axwalker** [reply]
If you want to have index renaming, I've been able to do this with `extendRoute`:

```js
extendRoute(route) {
  if (route.path.match(/\(.+\)/)) {
    route.parent.insert('index', route.components.get('default'))
    route.delete()
  }
}
```

`pages/users/[id]/(details).vue` works with this config though to match the `users/some-id/` route.

It should be noted this will only work for leaf routes, not intermediate folders. So you can't have a path that's something like `pages/(authorized)/users/[id]/(details).vue` with this.

**Reply by axwalker** [reply]
PS. @posva if you do happen to know how I could adjust my above snippet so that intermediate directories like `(authorized)` would just be mapped to `'index'` that would be amazing and open up a lot of options for me!

If you just try to have that directory in there you can get an error like:

```
Failed to resolve import "index" from "virtual:vue-router/auto-routes". Does the file exist?
virtual:vue-router/auto-routes:10:33
8  |          path: '',
9  |          name: '/example/',
10 |          component: () => import('index'),
   |                                  ^
11 |          /* no children */
12 |        }
```

**Reply by posva** [reply]
I don’t know from the top of my head. There is probably a way by combining the pages folder options and maybe the extend route. But maybe only using the options to parse folders is enough

**Comment by axwalker** [comment]
I'd also love to see route grouping like is available in SolidStart and SvelteKit ([their grouping docs](https://kit.svelte.dev/docs/advanced-routing#advanced-layouts-group)). I'd also love to be able to enforce directory only routes like SvelteKit opted for.

In the directory-only scenario, you would not be allow to have the following:

```md
- pages/
  - index.vue
  - films.vue
  - actors.vue
```

and instead would have to do:

```md
- pages/
  - index.vue
  - films/
    - index.vue
  - actors
    - index.vue
```

I realise this is probably not desirable as a default for vue-router's file based configuration, but it would be nice to have the option to enforce this. Having worked on several Vue, Nuxt, and SvelteKit projects I've personally found it a lot easier to mentally parse the routes/pages folders when I know a directory always maps to a route.

In the [linked SvelteKit docs]((https://kit.svelte.dev/docs/advanced-routing#advanced-layouts-group)) you'll see this also enables you to easily have different layouts for different groups. (I realise vue-router doesn't have quite the same concept of a 'layout' as SK's `+layout.svelte` stuff)

**Comment by posva** [comment]
FYI I added this to the board. I think the groups serves multiple purposes and it will also help with issues like

- #33 
- #318 
- #107

**Reply by robertmoura** [reply]
Thanks for that Eduardo!

For anyone interested, here is the item on the board: https://github.com/users/posva/projects/2/views/1?query=is%3Aopen+sort%3Aupdated-desc&pane=issue&itemId=68007322
